feat: Game captions & fix answer logic for synonyms and numbers

jschhie opened the pull request:
## $\color{Emerald}{Reference}$
- Issue [#257](https://github.com/Crustaly/audemywebsite/issues/257): Game Q&A captions
- Adds **accessibility & usability** for sighted users + devs
  - Especially for those w/ hearing or memory difficulties
  - Or for those in loud/quiet environments
- **Potentially reduces TTS API call**s by providing a visual alternative

- Related Issue [#97](https://github.com/Crustaly/audemywebsite/issues/97): Inconsistent number formats in user transcriptions

---
## $\color{Emerald}{Summary}$
### Key Changes
- Adds UI, logic, and props to `<GameHeader/>` to display dynamic captions for:
  - Game Q&A, feedback, and a progress bar
- Updates all games to pass the required props to `<GameHeader/>`
- Adds logic to handle answers with synonyms & numbers
  - **Before:** "Car Counting" game rejected **digit-based** numbers ('3') & only accepted **word-based** #'s ('three')
  - **After:** All Math games now account for both # formats
  - **After:** Game captions account for **synonyms** ('drink' vs 'beverage') & **plurality** ('drink' vs 'drinks')
- **Modified:** `validateAnswer()` & `validateCarAnswer()`
- Auto-Formats & adds RWD for multi-part Q's and multiple-choice Q's for UI readability

### Minor Updates
- **Refactor:** Adds/uses missing `gameConfigs.js` for "Car Counting"
- **Fix:** Replaces 40 duplicate **'chair'** answers in "Odd One Out" for better variety & gameplay UX
- **Fix:** Replaces blurry game icons on mobile

---
## $\color{Emerald}{ Game \ Header \ Updates}$
### 1. Feat: Add UI logic for dynamic captions & progress bar
<details>
  <summary>Details</summary>

### Key Props / Flags

| Flag | Desc |
| :--- | :--- |
| `showCaptions` | - Controls visibility for captions & progress bar<br> <br>**All 3 must be met:** <br> - Game is in progress <br>- Intro audio has finished <br>- Game has remaining Q's |
| `showAnswerOnly` | - Set to `true` for 'Spelling Bee' & 'Car Counting'<br> - Only show answer, progress bar, & game desc<br>- Hides captions for Qs |
| `isAnswerPlaying` & `isCorrect` | - Reactive, timed flags<br> - Returned from `validateAnswer()` |
| `firstMatchingAnswer` | - Helps align answer captions w/ transcript<br> - Accounts for synonyms & number formats<br>- Currently, applicable to all Math games & "Grocery Grab" |
| Other props not listed for brevity | --- |

</details>

---
## $\color{Emerald}{ Answer \ Validation \ Fixes }$
### 2. Fix: Answer Validation & Transcript Alignment for Synonyms & #'s
<details>
  <summary>Details</summary>

- **All Math games & "Car Counting"**:
  - Now accept either word-based ('three') or digit-based ('3') #'s
- **Grocery Grab:**
  - Answer captions for synonyms now align with the transcript
- **Affected:** `validateAnswer()` & `validateCarAnswer()`
- **Note:** All other games only accept a single answer per Q
  - But synonyms can be added in the future (via `firstMatchingAnswer`)

</details>

### 3. Feat: Auto-Format multi-part & multiple-choice Q's
<details>
  <summary>Details</summary>

### `splitMultiPartsQs()`
| `splitMultiPartsQs()` | Details |
| :--- | :--- |
| **Purpose** | - Splits multi-part Q's into a separate prompt and question |
| **Games Affected** | - Fruit Frenzy, Monkey Madness, Shape Shark |
| **Example Format** | - <kbd>'There are 3 apples and 2 bananas.'</kbd> → `prompt` <br/> - <kbd>'How many fruits are there in total?'</kbd> → `question` |

### `splitMCQs()`
| `splitMCQs()` | Details |
| :--- | :--- |
| **Purpose** | - Splits multiple-choice Q's into the actual question and a list of choices <br> - Handles logic for optional, trailing <kbd>"or "</kbd> |
| **Games Affected** | - Polar Pairing, Vocab Vortex, Odd One Out |
| **Example Format** | - <kbd>'Which word is the opposite of Freedom?'</kbd> → `question` <br/> - <kbd>'Independence, Constraint, Liberty, or Autonomy?'</kbd> → `choices` |

</details>

---
## $\color{Emerald}{ Screenshots }$
### Feat: Auto-Format Game Captions & Progress Bar (Demos)
> For brevity, most demos show RWD for `lg`-`2xl`

#### a. Language Games
<details>
  <summary>View</summary>

#### Fill-in-the-Blank Q's
| Fill-in-the-Blank Q | Answer/Feedback |
|:---:| :---:|
| <img width="500" alt="Screenshot 2025-09-07 at 1 03 14 PM" src="https://github.com/user-attachments/assets/23d4f963-0c26-472b-a7cb-27644c383073" /> |  <img width="500" alt="Screenshot 2025-09-07 at 1 03 18 PM" src="https://github.com/user-attachments/assets/ffb23265-3a0c-4edd-a656-a094ab9b7ee9" /> |

#### Special Format + RWD: Multiple Choice Q's
| mobile & small | medium | large, xl, 2xl |
|:---:|:---:|:---:|
| <img width="300" alt="MCQ answer lang mobile small" src="https://github.com/user-attachments/assets/2becd3b1-7925-4671-92df-27b223ebd103" /> | <img width="450" alt="MCQ lang medium" src="https://github.com/user-attachments/assets/45894bde-3330-4401-b219-5321cd099ffc" /> | <img width="500" alt="MCQ lang lg xl 2xl" src="https://github.com/user-attachments/assets/0d32a4f6-3812-4e04-80cb-8fd146cfdf02" /> |

</details>

#### b. Math Games
<details>
  <summary>View</summary>

| Fill-in-the-Blank Q's | Multi-Parts Q's (prompt + Q) |
| :---:| :---:|
| <img width="500" alt="FIB math lg xl 2xl reply" src="https://github.com/user-attachments/assets/18ef2d25-58b5-4603-96fa-130a893358a8" /> | <img width="500" alt="multipart lg xl 2xl reply" src="https://github.com/user-attachments/assets/365fcc93-77b3-4eca-a9b0-1a0ed4c27ee2" /> |

</details>

#### c. Spelling Bee & Car Counting Only
<details>
  <summary>View special cases</summary>

> - Display only Answers, Progress Bar, & Game Desc. (hide Question captions) 
> - Car Counting (& Math games): Accepts both digit-based & word-based #'s

| Spelling Bee | Car Counting |
| :---:| :---:|
| <img width="500" alt="lg xl 2xl reply" src="https://github.com/user-attachments/assets/7fc39d9d-466c-4793-bbc4-8ee3e81ea4a4" /> | <img width="500" alt="digit format answers car counting" src="https://github.com/user-attachments/assets/dfa8bbf3-ddc5-41ea-a85b-e87a4b666f24" /> |

</details>

#### d. Other Game Categories (All Fill-in-the-Blank Q's)
<details>
  <summary>View</summary>

| Science | Life Skills | Independence Skills |
| :---: | :---:| :---:|
| <img width="400" alt="science wrong" src="https://github.com/user-attachments/assets/a69e1f8d-7ba3-4d3d-a396-4e71d1650af2" /> | <img width="400" alt="life skills correct" src="https://github.com/user-attachments/assets/0cbfbab1-0069-4392-bfbd-1470cdce70eb" /> | <img width="400" alt="indep skills correct" src="https://github.com/user-attachments/assets/b3ac0c21-b64b-48a7-8fc3-9867db075a10" /> |

</details>

---
### Fix: Answer Validation & Transcript Alignment for Synonyms & #'s (Demos)
#### a. Grocery Grab: Synonyms
<details><summary>View</summary>

> (Sample Q&A)
> - **Q:** "Which aisle can you find orange juice?"
> - **A:** ["beverage", "beverages", "drink", "drinks"]

| Synonym 1 | Synonym 2 |
| :---: | :---: |
|<img width="500" alt="grocery orange juice - synonyms 1" src="https://github.com/user-attachments/assets/0fa7ef3d-97c8-45a5-9b11-ccc6f4b6c0f5" /> | <img width="500" alt="grocery orange juice - synonyms 2" src="https://github.com/user-attachments/assets/2cc29026-6eaf-42c6-afc1-466d7dbfb014" /> |

</details>

#### b. Car Counting: Number Formats
<details><summary>View</summary>

| Word-based Transcript/Captions | Digit-based Transcript/Captions |
| :---: | :---: |
| <img width="500" alt="lg xl 2xl answer" src="https://github.com/user-attachments/assets/d04d7660-06a7-4c70-bda8-a74d2c05dbcb" /> | <img width="500" alt="digit format answers car counting" src="https://github.com/user-attachments/assets/dfa8bbf3-ddc5-41ea-a85b-e87a4b666f24" />|

</details>

---
## $\color{Emerald}{ Manual \ Tests}$
<details>
  <summary>View</summary>

- All games were individually tested:
  - **Game play:** Start Q, Repeat Q, Record/Submit Answer, transcription, captions, feedback, & scoring
  - **RWD:** Screen widths b/t `375px` & `1680px`
  - **Number Formats**: Both digit-based & word-based #'s are accepted in all Math games 
  - **Synonyms:** Transcript/captions alignment for answers with synonyms

</details>

---
## $\color{Emerald}{ Future \ Work}$
<details><summary>View todo</summary>

- Fix `validateAnswer()` to handle answers with **whitespace**
  - **Ex:** "right triangle" _(Shape Shark)_, or "white and black" _(Color Game)_
  - **Current logic (`preprod`):**
    - Only handles single words _(ex: "square", "red")_
    - & Erroneously marks correct transcriptions
- Idea: Convert UI for MCQ answers → functional <kbd>buttons</kbd>

</details>

jschhie committed:
feat: Add game question captions in <GameHeader/>

- Add props/logic to <GameHeader/> to display accessible Qs & current index
- Use data from 'useGameCore.js' for dynamic & async updates

jschhie committed:
feat: Captions for Definition Detective game Qs

jschhie committed:
fix: RWD for question captions & style UI for <GameHeader/>

jschhie committed:
feat: Captions for non-MCQ Language games

- Update <GameHeader/> props passed in non-multiple choice Language games
- Games affected:
    - Part of Speech, Color Game, & Syllable Sorting

jschhie committed:
feat: Auto-format multi-choice Qs in <GameHeader/>

- Add 'multipleChoiceGames' array for MCQ games (ex: Vocab Vortex)

jschhie committed:
feat: Captions for Vocab Vortex w/ MCQ formatting

jschhie committed:
feat: Handle trailing 'or' in MCQs (<GameHeader/>)

- Update splitMCQs() to format/trim 'or' before last MC option (edge case)
- Adjust RWD flex for MCQs

jschhie committed:
feat: Captions for Polar Pairing w/ MCQ formatting

- NOTE: <GameHeader/> formats trailing 'or' in MC options (parent commit)

jschhie committed:
feat: Captions for Odd One Out w/ MCQ formatting

- Add game to 'multipleChoiceGames' & adjust RWD in <GameHeader/>

jschhie committed:
fix: Correct 'currentQuestion' prop in <GameHeader/>

- Fix prop type (string --> Object) & set default value (null)
- Add null check to display 'currentQuestion'

jschhie committed:
feat: Add question #s for Spelling Bee captions

- Add conditional logic to show current game Qn index & desc in <GameHeader/>

jschhie committed:
refactor: mobile RWD for icons & text in <GameHeader />

- Improve RWD to reduce scrolling on mobile/small screens
- Minor updates: Hide icons from SR output & add RWD transitions

jschhie committed:
feat: Captions for Math games Qs

- Update <GameHeader/> props passed for each Math game
- Games affected:
    - Fruit Frenzy, Shape Shark, Animal Add.,
    - Sub. Safari, Multi. Madness,
    - Division Duel, & Monkey Madness

jschhie committed:
feat: Add question #s for Car Counting captions

jschhie committed:
feat: Captions for Science games Qs

- Update <GameHeader/> props passed in each Science game
- Affected: All 10 Science games

jschhie committed:
feat: Captions for Life Skills games Qs

- Update <GameHeader/> props passed in each Life Skill game
- Affected: All 14 Life Skill games

jschhie committed:
feat: Captions for Independence Skills games Qs

- Update <GameHeader/> props passed in each Indep. Skills game
- Affected: All 12 Indep. Skills games

jschhie committed:
feat: Logic/UI for multi-part games captions in <GameHeader/>

- Add splitMultiPartsQs(): Splits prompt from Q in <GameHeader/> for UI readability
- Games affected: Fruit Frenzy, Monkey Madness, & Shape Shark
- Minor refactor: Rename variable in splitMCQs() for clarity ('prompt' --> 'question')

jschhie committed:
fix: Replace blurry game icons & remove previous ones from repo

jschhie committed:
fix: transcription RWD & legibility in <GameControls/>

- Add transparent bg; update UI for better RWD & style
- Re-run Prettier

jschhie committed:
feat: Answer captions & progress bar in <GameHeader/>

Update useGameCore.js:
- Add reactive, timed 'isAnswerPlaying' & 'isCorrect' flags
- Add new flags to return values

Update <GameHeader/>:
- Pass matching props for dynamic answer captions
- Add UI logic for question progress bar
- Rename 'showQuestions' prop to 'showCaptions' for scope/clarity

Minor Refactor:
- Update RWD for new feats
- Extract classes for DRY code & readability; add inline docs

jschhie committed:
feat: Answer captions for Language Games

- Pass new props to <GameHeader/> for Fill-in-the-Blank Lang. Games

Affected:
- All Language Games (MCQ & Fill-in-the-Blank Qns)
- (Except Spelling Bee; handled separately)

jschhie committed:
feat: Answer captions for Spelling Bee

- Add 'showAnswerOnly' prop in <GameHeader/> to hide Qns
- Flag affects: Spelling Bee & Car Counting games
- Pass matching props to <GameHeader/> for Spelling Bee

jschhie committed:
feat: Answer captions for Math games

- Pass matching props to <GameHeader/>
- Affected: All Math Games (Multi-part & Fill-in-the-Blank Qns)
- (Except Car Counting; handled separately)

jschhie committed:
feat: Answer/feedback captions for Car Counting

- Add reactive, timed 'isCorrect' & 'isAnswerPlaying' flags in CarCounting.vue
- Pass props to <GameHeader/>
- Add UI logic for Car Counting answer captions in <GameHeader/>

jschhie committed:
feat: Answer/feedback captions for Science games

- Pass props to <GameHeader/> for all 10 Science games

jschhie committed:
feat: Answer/feedback captions for Life Skills games

- Pass props to <GameHeader/> for all 14 Life Skills games

jschhie committed:
fix: Align <GameHeader/> captions for synonyms & #s

- Align answer captions w/ transcription for synonyms and # formats
- Add 'foundWord' return value to validateAnswer() for matching
- Implement 'firstCorrectAnswer' prop & UI logic in <GameHeader/>
- Affected: All Math games (except Car Counting) & Grocery Grab
- Minor: 2xl RWD for captions

jschhie committed:
fix: Answers & captions w/ digit-based numbers for Car Counting

- Add reactive 'isWordBasedFormat' flag & 'wordToDigitMap' in useCarCounting.js
- Fix validateCarAnswer() to check if transcript has digits
- Fix getCurrentAnswer() to align answer captions w/ transcript word/digit format
- Rename 'answerMap' to 'digitToWordMap' for clarity
- Minor: Refactor inline <GameHeader/> classes into constants

jschhie committed:
refactor: Add/Use gameConfigs.js for Car Counting

- Add missing game entry in gameConfigs.js
- Replace hardcoded values w/ gameConfigs in carCounting.vue

jschhie committed:
feat: Answer/feedback captions for Indep. Skills games

- Pass props to <GameHeader/> for all 12 Independence Skills games

jschhie committed:
fix: 40 duplicate Odd One Out 'chair' answers

- Replace 40 duplicate 'chair' answers with alternative choices
- Improve QnA variety for better gameplay/UX

master-of-none commented:
@jschhie 
Is this ready for review?

jschhie commented:
Yes, it's ready for review! 

Thank you for asking, @master-of-none.

The PR desc is a bit detailed, but the "Summary", "Screenshots", & "Manual Tests" should provide the key points.

master-of-none commented:
> Yes, it's ready for review!
> 
> Thank you for asking, @master-of-none.
> 
> The PR desc is a bit detailed, but the "Summary", "Screenshots", & "Manual Tests" should provide the key points.

Yes. I will review it in detail. Will take a day or two to review, test and approve it. 
Thanks !

jschhie commented:
Thanks again & no rush! @master-of-none

jschhie committed:
refactor: Rename 'firstCorrectAnswer' to 'firstMatchingAnswer'

- Improve variable clarity; no logical or UI changes

NOTES:
- Code finds 1st matching answer to transcript, not 1st answer in array
- Variable prevents UI mismatch in answer captions for <GameHeader/>

jschhie commented:
@master-of-none 

**Update:** Just pushed one final commit to rename `firstCorrectAnswer` to `firstMatchingAnswer` for variable clarity

(No logical changes made)

master-of-none commented on public/assets/questionsDb/oddOneOutDB.json:
```diff
@@ -3,8 +3,8 @@
-          "Q": "Which word is the odd one out? fish, dog, chair, cat",
-          "A": ["chair"]
+          "Q": "Which word is the odd one out? fish, dog, moon, cat",
+          "A": ["moon"]
```
Why are the questions changed @jschhie

jschhie replied:
@master-of-none 

I noticed that 'Odd One Out' game had 40 duplicate 'chair' answers, which made gameplay a bit repetitive. 

So I updated its Q&A to use different words for better variety/UX (This minor change was also noted in the PR desc).

master-of-none commented:
# PR Review – Game Q&A Captions & Answer Validation Fixes

## Overall Impression
This is a **substantial improvement** for accessibility, usability, and gameplay UX.  
The addition of captions, synonym/number handling, and responsive formatting directly addresses multiple open issues (#257, #97).  
The changes appear well-tested across games and devices, and the demos/screenshots are very clear.

---

##  Strengths
- **Accessibility & Inclusivity**
  - Captions improve support for users with hearing/memory difficulties and in noisy environments.
  - Visual feedback reduces reliance on TTS.

- **Answer Validation**
  - Synonym + plural handling (`drink` vs `drinks`).
  - Number format handling (`3` vs `three`).
  - Consistent logic across games.

- **UI/UX**
  - Dynamic captions + progress bar in `<GameHeader/>`.
  - Auto-formatting for multi-part and multiple-choice Q’s improves readability.
  - Replaced blurry icons & duplicate answers for smoother gameplay.

- **Code Quality**
  - Added `gameConfigs.js` → improves maintainability.
  - Centralized validation logic reduces inconsistencies.

- **Testing**
  - Manual tests across screen widths (375px–1680px).
  - Covered multiple game categories & answer types.

---

## Areas for Improvement
1. **Whitespace Handling (Future Work)**
   - Currently, multi-word answers like `"right triangle"` aren’t handled.
   - Suggest prioritizing this fix in the next PR since it directly affects correctness.

2. **Validation Logic**
   - `validateAnswer()` and `validateCarAnswer()` could be further unified to reduce duplication.
   - Consider extracting a shared `normalizeAnswer()` utility (handles lowercase, trimming, plural, synonyms, digits).

3. **Component Responsibility**
   - `<GameHeader/>` is taking on captions, progress bar, and formatting logic.
   - Might be worth splitting into smaller subcomponents (`<GameCaption/>`, `<GameProgress/>`) for clarity and reusability.

4. **Testing Coverage**
   - Manual tests are great, but automated unit tests for:
     - Synonym mapping
     - Digit/word number handling
     - RWD rendering
   - would help prevent regressions.

---

## Suggestions / Future Work
- Convert MCQ UI from plain captions → **interactive buttons** (already noted).
- Add automated **a11y checks** (e.g., contrast, ARIA roles for captions).
- Consider **i18n/l10n support** for synonyms and number formats (multi-language users).
- Improve developer documentation around `gameConfigs.js` and validation rules.

---

## Verdict
This PR makes **big strides in accessibility and consistency**.  
Merging is recommended, with follow-up issues for whitespace handling and potential refactor into smaller components.  

Great job on the thorough testing and detailed screenshots!

jschhie commented:
Thanks for the review, @master-of-none! 

I saw your comment about the “Odd One Out” Q&As and left a [reply](https://github.com/Crustaly/audemywebsite/pull/268#discussion_r2350554248).

Once this PR is merged, I'll create a follow-up to address your suggestions, including Issue #270.

master-of-none reviewed (approved):
As mentioned in https://github.com/Crustaly/audemywebsite/pull/268#issuecomment-3294678001

Approved !
